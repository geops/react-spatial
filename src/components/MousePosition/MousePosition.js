import React, { useState, useRef, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import OLMap from 'ol/Map';
import { createStringXY } from 'ol/coordinate';
import OLMousePosition from 'ol/control/MousePosition';

const propTypes = {
  /**
   * The current map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * List of projections to display.
   */
  projections: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The label to display in the select box.
       */
      label: PropTypes.string.isRequired,

      /**
       * The value used to create the options´s projection of the MousePosition control.
       * See [doc](https://openlayers.org/en/latest/apidoc/module-ol_control_MousePosition.html).
       */
      value: PropTypes.string.isRequired,

      /**
       * A function following the  [CoordinateFormat](https://openlayers.org/en/latest/apidoc/module-ol_coordinate.html#~CoordinateFormat).
       */
      format: PropTypes.func,
    }),
  ),

  /**
   * The initially selected projection
   */
  projectionValue: PropTypes.shape({
    /**
     * The label to display in the select box.
     */
    label: PropTypes.string.isRequired,

    /**
     * The value used to create the options´s projection of the MousePosition control.
     * See [doc](https://openlayers.org/en/latest/apidoc/module-ol_control_MousePosition.html).
     */
    value: PropTypes.string.isRequired,

    /**
     * A function following the  [CoordinateFormat](https://openlayers.org/en/latest/apidoc/module-ol_coordinate.html#~CoordinateFormat).
     */
    format: PropTypes.func,
  }),

  /**
   * Function triggered on projection's change event.
   * @param {Event} event The change event object.
   * @param {Object} projection The selected projection object.
   */
  onChange: PropTypes.func,
};

const defaultProps = {
  onChange: () => {},
  projections: [
    {
      label: 'EPSG:4326',
      value: 'EPSG:4326',
    },
    {
      label: 'EPSG:3857',
      value: 'EPSG:3857',
    },
  ],
  projectionValue: null,
};

function MousePosition({
  map,
  projections,
  projectionValue,
  onChange,
  ...other
}) {
  const [projection, setProjection] = useState(
    projections &&
      ((projectionValue &&
        projections.find((p) => p.value === projectionValue.value)) ||
        projections[0]),
  );
  const [control, setControl] = useState();
  const ref = useRef();

  useEffect(() => {
    const mousePosition = new OLMousePosition({
      target: ref.current,
      undefinedHTML: '&nbsp;',
      className: '',
    });
    map.addControl(mousePosition);
    setControl(mousePosition);

    return () => {
      map.removeControl(mousePosition);
    };
  }, []);

  useEffect(() => {
    if (!projection || !control) {
      return;
    }
    control.setProjection(projection.value);
    control.setCoordinateFormat(projection.format || createStringXY(4));
  }, [projection, control]);

  const onChangeCb = useCallback(
    (evt) => {
      const newProj = projections.find((opt) => evt.target.value === opt.value);
      setProjection(newProj);
      onChange(evt, newProj);
    },
    [projections],
  );

  if (!projection || !projections || !projections.length) {
    return null;
  }

  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <div className="rs-mouse-position" {...other}>
      <select
        className="rs-select"
        value={projection.value}
        onChange={onChangeCb}
      >
        {projections.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
      <span ref={ref} className="rs-coordinates" />
    </div>
  );
}

MousePosition.propTypes = propTypes;
MousePosition.defaultProps = defaultProps;

export default React.memo(MousePosition);
