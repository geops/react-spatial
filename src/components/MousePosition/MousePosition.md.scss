.rs-mouse-position-example {
  .rs-mouse-position {
    display: flex;
    align-items: center;
    margin: 10px 0;

    span {
      margin-left: 10px;
    }

    select {
      background: none;
      height: 20px;
    }
  }
}
