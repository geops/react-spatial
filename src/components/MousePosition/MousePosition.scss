.tm-mouse-position {
  & > select {
    display: inline-block;
  }
}
