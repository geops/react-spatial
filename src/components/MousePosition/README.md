#

This demonstrates the use of MousePosition.

```jsx
import React from  'react';
import OLMap from 'ol/Map';
import BasicMap from 'react-spatial/components/BasicMap';
import LayerService from 'react-spatial/LayerService';
import ConfigReader from 'react-spatial/ConfigReader';
import MousePosition from 'react-spatial/components/MousePosition';

class MousePositionExample extends React.Component {
  constructor(props) {
    super(props);
     this.center = [1149722.7037660484, 6618091.313553318];
    this.map = new OLMap({controls:[]});
    this.state = {
      layerService: null
    }
  }

  componentDidMount() {
    const layers = ConfigReader.readConfig(
      this.map,
      treeData,
    );
    this.setState({
      layerService: new LayerService(layers)
    })
  }

  render() {
    return (
      <div className="tm-mouse-position-example">
        <BasicMap map={this.map} center={this.center} zoom={6} />
        <MousePosition map={this.map} projections={[
        {
          label: 'World Geodetic System 1984',
          value: 'EPSG:4326',
        },
        {
          label: 'Web mercator with custom output',
          value: 'EPSG:3857',
          format: (coord) => {
            const decimals = 4;
            let text = [];

            coord.forEach((input)=> {
                input = Math.round(parseFloat(input) * Math.pow(10, decimals)) /
                    Math.pow(10, decimals);
                input =input.toString();
                var parts = input.toString().split('.');
                parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, "'");
                text.push(parts.join());
            });
            return 'Coordinates: ' + text.join(' ');
          }
        },]}/>
      </div>
    )
  }
}

<MousePositionExample />;
```
