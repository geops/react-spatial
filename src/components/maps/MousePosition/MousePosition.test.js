/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-canvas-mock';
import OLMap from 'ol/Map';
import OLMousePosition from 'ol/control/MousePosition';
import MousePosition from './MousePosition';

configure({ adapter: new Adapter() });
const expectSnapshot = (props) => {
  const map = new OLMap({});
  const component = renderer.create(<MousePosition map={map} {...props} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
};

describe('MousePosition', () => {
  describe('matches snapshots', () => {
    test('using default values.', () => {
      expectSnapshot({});
    });

    test('using no projections.', () => {
      expectSnapshot({
        projections: [],
      });
    });

    test('using only one projection', () => {
      expectSnapshot({
        projections: [{ label: 'foo', value: 'foo' }],
      });
    });

    test('using multiple projections.', () => {
      expectSnapshot({
        projections: [
          { label: 'foo', value: 'foo' },
          { label: 'bar', value: 'bar' },
        ],
      });
    });
  });

  test('add MousePosition control to the map.', () => {
    const map = new OLMap({});
    const spy = jest.spyOn(map, 'removeControl');
    const spy2 = jest.spyOn(map, 'addControl');
    const fn = jest.fn();
    mount(
      <MousePosition
        map={map}
        projections={[
          {
            label: 'EPSG:4326',
            value: 'EPSG:4326',
            format: fn,
          },
        ]}
      />,
    );
    const ctrl = spy2.mock.calls[0][0];
    expect(spy).toHaveBeenCalledTimes(0);
    expect(spy2).toHaveBeenCalledTimes(1);
    expect(ctrl).toBeInstanceOf(OLMousePosition);
    expect(ctrl.getProjection().getCode()).toBe('EPSG:4326');
    expect(ctrl.getCoordinateFormat()).toBe(fn);
  });

  test('add/remove MousePosition control on mount/unmount.', () => {
    const map = new OLMap({});
    const spy = jest.spyOn(map, 'removeControl');
    const spy2 = jest.spyOn(map, 'addControl');
    const wrapper = mount(<MousePosition map={map} />);
    expect(spy).toHaveBeenCalledTimes(0);
    expect(spy2).toHaveBeenCalledTimes(1);
    wrapper.unmount();
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0]).toBe(spy2.mock.calls[0][0]);
  });

  test('triggers onChange when select projection.', () => {
    const map = new OLMap({});
    const onChange = jest.fn(() => {});
    const wrapper = mount(
      <MousePosition
        map={map}
        onChange={onChange}
        projections={[
          {
            label: 'EPSG:4326',
            value: 'EPSG:4326',
            format: jest.fn(),
          },
        ]}
      />,
    );
    // onChange triggered on instantiation.
    expect(onChange).toHaveBeenCalledTimes(0);
    wrapper.find('select').simulate('change', {});
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('applies new format and value when we select a new projection.', () => {
    const map = new OLMap({});
    const spy = jest.spyOn(map, 'addControl');
    const projs = [
      {
        label: 'EPSG:4326',
        value: 'EPSG:4326',
        format: jest.fn(),
      },
      {
        label: 'EPSG:3857',
        value: 'EPSG:3857',
        format: jest.fn(),
      },
    ];
    const wrapper = mount(<MousePosition map={map} projections={projs} />);

    const ctrl = spy.mock.calls[0][0];
    expect(ctrl.getProjection().getCode()).toBe(projs[0].value);
    expect(ctrl.getCoordinateFormat()).toBe(projs[0].format);
    wrapper
      .find('select')
      .simulate('change', { target: { value: 'EPSG:3857' } });
    expect(ctrl.getProjection().getCode()).toBe(projs[1].value);
    expect(ctrl.getCoordinateFormat()).toBe(projs[1].format);
  });
});
