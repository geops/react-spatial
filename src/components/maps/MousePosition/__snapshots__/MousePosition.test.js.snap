// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MousePosition matches snapshots using default values. 1`] = `
<div
  className="rs-mouse-position"
>
  <select
    className="rs-select"
    onChange={[Function]}
    value="EPSG:4326"
  >
    <option
      value="EPSG:4326"
    >
      EPSG:4326
    </option>
    <option
      value="EPSG:3857"
    >
      EPSG:3857
    </option>
  </select>
  <span
    className="rs-coordinates"
  />
</div>
`;

exports[`MousePosition matches snapshots using multiple projections. 1`] = `
<div
  className="rs-mouse-position"
>
  <select
    className="rs-select"
    onChange={[Function]}
    value="foo"
  >
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
  </select>
  <span
    className="rs-coordinates"
  />
</div>
`;

exports[`MousePosition matches snapshots using no projections. 1`] = `null`;

exports[`MousePosition matches snapshots using only one projection 1`] = `
<div
  className="rs-mouse-position"
>
  <select
    className="rs-select"
    onChange={[Function]}
    value="foo"
  >
    <option
      value="foo"
    >
      foo
    </option>
  </select>
  <span
    className="rs-coordinates"
  />
</div>
`;
